src/util/rust_util.rs

```diff
@@ -6,3 +6,71 @@ pub const fn min_of_usize(a: usize, b: usize) -> usize {
         a
     }
 }
+
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+/// Increment the atomic number by 1. If the number reaches limit, return true and reset the number.
+/// Otherwise, return false. This function is thread safe so only the first call that makes the number
+/// reach limit will return true.
+// clippy suggests to use match new_val.cmp(limit) { Ordering::Greater => ..., Ordering::Equal => ..., Ordering::Less => ... },
+// which is unnecessary. Plus clippy lists a known issue with match/ordering that it is slower than if.
+#[allow(clippy::comparison_chain)]
+pub fn increment_and_check_limit(u: &AtomicUsize, limit: usize) -> bool {
+    match u.fetch_update(Ordering::SeqCst, Ordering::Relaxed, |current| {
+        // New value after increment
+        let new_val = current + 1;
+        if new_val < limit {
+            // If the new value is smaller than limit, set to the new value.
+            Some(new_val)
+        } else if new_val == limit {
+            // If the new value equals the limit, reset to 0
+            Some(0)
+        } else {
+            // If the new value is larger than the limit, this should not happen as
+            // we increment by 1 and we should have reset it to zero when the value ever reaches the limit.
+            unreachable!()
+        }
+    }) {
+        // We are not returning None from the update func, so Err is unreachable.
+        Err(_) => unreachable!(),
+        Ok(old) => old + 1 == limit,
+    }
+}
```

Comment:
What about letting it stick to the maximum value?
```suggestion
fn increment_and_check_limit(u: &AtomicUsize, limit: usize) -> bool {
    let old = u.fetch_add(1, Ordering::SeqCst);
    if old >= limit {
        panic!("Incremented past limit: {} -> {}, limit: {}", old, old + 1, limit);
    }
    old + 1 == limit
}
```

Reply:
With this change the code would be cleaner. However, one implication from the change is that we would always need some sort of global synchronization (to reset `scanned_stacks`) before running any `ScanStackRoot` packet. Currently we have no such requirement for `ScanStackRoot` -- we can just schedule them and they will work. I am not confident that the current implementation is correct though (especially whether it is correct if the number of mutators changes somehow). I will think about this a little bit more.

Reply:
I think it is better to have such "global synchronization".  The sole purpose of shared atomic counter is synchronization.  Stack scanning is a typical fork-join use pattern. After stopping the world, we spawn N work packets for N stacks, and wait for those N packets to complete.  I think we effectively create a "goal" of "do something after N stacks are scanned" the moment we spawn N work packets, so that's a natural point to initialize the counter, and it is why we have the counter in the first place.  It is also a kind of "global synchronization" that we "do things (shift the GC phase) **after** the counter reaches N".

Speaking of reading/writing shared variable, in Java, it is also natural to write to a variable before forking (`Thread.start`), and read from the variable in sub-threads.  Although we did not use any locks for synchronization, the Java memory model still considers the write as happening-before the read (`Thread.start` happens before the first operation of the started `Thread`), and is still correctly synchronized.

Speaking of initializing the counter, I think we can also "destroy" the counter when N stacks have been scanned.  By "destroying" I mean any subsequent increment of the counter shall result in error.  Probably we can invent our own `CountDownLatch` data structure that dictates a particular use pattern, i.e. `new()`, `set_what_to_do_when_counter_reached_limit(callback)`, `increment()`, `reset()`.  A key difference is that such `CountDownLatch` does not block a thread until the latch is released, but does something (or enables a work packet) when the count reached limit.

Reply:
I have made changes. As with this change, the function `increment_and_check_limit()` is simple enough, and I don't think we need it to be a separate function any more. I have deleted `increment_and_check_limit()`. I put the code to `BasePlan`. You can take a look now.